File: cpp/2016.maximum-difference-between-increasing-elements.cpp
#include <vector>
using std::vector;

class Solution {
public:
  int maximumDifference(vector<int> &nums) {
    int maxDiff = -1, mmin = std::numeric_limits<int>::max();

    for (int num : nums) {
      mmin = std::min(mmin, num);
      maxDiff = std::max(maxDiff, num - mmin);
    }

    return maxDiff > 0 ? maxDiff : -1;
  }
};

File: golang/2016.maximum-difference-between-increasing-elements.go
func maximumDifference(nums []int) int {
	maxDiff := -1
	curMin := nums[0]

	for _, num := range nums {
		curMin = min(curMin, num)
		maxDiff = max(maxDiff, num-curMin)
	}

	if maxDiff == 0 {
		return -1
	}

	return maxDiff
}
